_drafts/printf-on-embedded.md

```diff
@@ -0,0 +1,566 @@
+`_write` will be called for every byte. You might want to instead buffer up to
+line ends, by:
+
+```c
+#include <stdio.h>
+setvbuf(stout, NULL, _IOLBF, 0);
+```
+
+See `man setvbuf` or <http://www.cplusplus.com/reference/cstdio/setvbuf/> for
+details.
+
+### Reentrancy
+
+Fancy word for if a function can be safely premepted and called simultaneously
+from multiple (interrupt) contexts.
+
+Some libc functions in newlib share a context structure that contains global
+data structures. If you want to safely use them in multiple threads, you have a
+couple of options:
+
+- only use these functions in one thread in your program
+- disable interrupts when using non-reentrant functions
+- only use the reentrant versions of the functions (denoted by `_xxx_r()`, eg
+  `_printf_r` in newlib)
+- recompile newlib with `__DYNAMIC_REENT__` to have the non-reentrant functions
+  call a user-defined function to get the current thread's reentrancy structure
+- override the default global reentrancy structure as necessary
+
+See the header file here for some detailed information:
+
+<https://github.com/bminor/newlib/blob/6226bad0eafe762b811c62d1dc096bc0858b0d1a/newlib/libc/include/reent.h>
+
+And great explanations here- especially if using FreeRTOS, there's built-in
+support for managing reentrancy structures:
+
+- <https://www.codeinsideout.com/blog/freertos/reentrant/>
+- <https://mcuoneclipse.com/2020/11/15/steps-to-use-freertos-with-newlib-reentrant-memory-allocation/>
+
+Let's go on a small side adventure into reentrancy approaches for
+newlib\[-nano\]!
+
+#### One thread only
+
+It might not be necessary to use the non-reentrant functions in small interrupt
+handlers (and is probably a reasonable idea to prohibit them).
+
+Here's an approach that will cause a runtime error if a non-reentrant function
+is called in an interrupt context. Unfortunately it requires rebuilding newlib,
+so it's really only for curiousity or the very paranoid.
+
+First, rebuild newlib (or build it as part of your project build; this is
+probably really only practical if you're using a compiler cache like ccache),
+setting `__DYNAMIC_REENT__` when building.
+
+Then, when building your application source, define a `__getreent()` function
+with the following contents:
+
+```c
+#include <reent.h>
+// Function to check if in interrupt context
+int isInInterruptContext(void) {
+    // Read the CONTROL register
+    uint32_t control_reg;
+    asm volatile ("MRS %0, CONTROL" : "=r" (control_reg));
+
+    // Check the nPRIV bit (bit 0)
+    return (control_reg & 1) != 0;
+}
+struct _reent * __getreent(void) {
+  if (isInInterruptContext()) {
+    __asm__ ("bkpt 0");
+  }
+  // return the built in global reentrancy context
+  return _impure_ptr;
+}
+```
+
+Also be sure to also define `__DYNAMIC_REENT__` in your CFLAGS when building
+your application source.
+
+This will trip a software breakpoint if a non-reentrant function is called from
+an interrupt context!
```

Comment:
I'm not sure I follow the value of this. Not sure if it's my brain being smooth, but why would we want to trigger a software breakpoint here? Would this just be for local debugging to figure out if we were using these functions in a re-entrant way?

Reply:
oh yeah good call, will tidy this up- the idea is to crash if using a non-reentrant function from isr